File: face_detection_window.py
#! /usr/bin/env python

import cv2
import time
from collections import deque
import detector
import sys

def draw_stats(im, stats):
    queue, queue_sum = stats['queue'], stats['queue_sum']
    for i, s in enumerate([
            'fps: %.02f'%(queue[-1][0]),
            'min: %.02f'%(min(x[0] for x in queue)),
            'max: %.02f'%(max(x[0] for x in queue)),
            'sum: %.02f'%(queue_sum),
            'len: %d'%(len(queue))
    ]):
        cv2.putText(
            im, s, (0,15*(1+i)),
            cv2.FONT_HERSHEY_PLAIN, 1, (255,)*3
        )

def update_stats(stats, dt):
    queue, queue_sum = stats['queue'], stats['queue_sum']
    fps = 1/dt
    queue.append((fps, dt))
    queue_sum += dt
    while queue_sum > 1 and len(queue) > 1:
        fps, dt = queue.popleft()
        queue_sum -= dt
    stats['queue_sum'] = queue_sum



def detect(image, cascade, config):
    for rect in detector.detect(image, cascade, config):
        x, y, w, h = rect
        p1 = (x, y)
        p2 = (x+w, y+h)
        cv2.rectangle(image, p1, p2, (255, 0, 0), 1)

def func(config):
    window_name = 'window'
    cap = cv2.VideoCapture()
    if config['stream'] is None:
        cap.open(0)
    else:
        cap.open(config['stream'])

    cv2.namedWindow(window_name)
    try:
        tm = time.time()
        stats = {
            'queue': deque(),
            'queue_sum': 0
        }
        cascade = detector.make_detector(config)
        while True:
            dt = time.time() - tm
            tm += dt
            if dt > 1/10:
                for _ in range(int(dt / (1/10))):
                    cap.grab()
            ret, frame = cap.read()
            if not ret:
                break
            update_stats(stats, dt)
            draw_stats(frame, stats)
            detect(frame, cascade, config)
            cv2.imshow(window_name, frame)
            if cv2.waitKey(20) == ord(' '):
                break
    finally:
        cap.release()
        cv2.destroyWindow(window_name)

if __name__ == '__main__':
    config = detector.config(sys.argv[1:])
    func(config)

File: face_detection_window.cpp
#include <opencv2/opencv.hpp>
#include <opencv2/core/ocl.hpp>
#include <iostream>
#include <chrono>
#include <string>
#include <deque>
#include <vector>
#include <sstream>
#include <algorithm>
#include "detector.h"

using namespace std;
using namespace cv;

#define range(x) begin(x), end(x)

struct Stats{
    struct Rec{
        float fps, dt;
        Rec(float fps, float dt) : fps(fps), dt(dt) {}
    };
    float sum;
    deque<Rec> queue;
};

void update_stats(Stats& stats, float dt){
    float fps = 1 / dt;
    stats.queue.emplace_back(fps, dt);
    stats.sum += dt;
    while(stats.sum > 1. && stats.queue.size() > 1){
        Stats::Rec rec = stats.queue[0];
        stats.queue.pop_front();
        stats.sum -= rec.dt;
    }
}

void draw_stats(Stats const & stats, Mat image){
    vector<string> strs;
    ostringstream os;
    os << fixed << setprecision(2);
    auto cmp = [](Stats::Rec const & a, Stats::Rec const & b){
        return a.fps < b.fps;
    };

    os << "fps: " << stats.queue[stats.queue.size()-1].fps;
    strs.push_back(os.str()); os.str(string());
    os << "min: " << (*min_element(range(stats.queue), cmp)).fps;
    strs.push_back(os.str()); os.str(string());
    os << "max: " << (*max_element(range(stats.queue), cmp)).fps;
    strs.push_back(os.str()); os.str(string());
    os << "sum: " << stats.sum;
    strs.push_back(os.str()); os.str(string());
    os << "len: " << stats.queue.size();
    strs.push_back(os.str()); os.str(string());

    for(int i = 0; i < strs.size(); ++i){
        string s = strs[i];
        putText(
            image, s, Point(0, (i+1)*15),
            FONT_HERSHEY_PLAIN, 1, Scalar(255, 255, 255)
        );
    }
}

void detect(Mat& image, CascadeClassifier& detector, CascadeConfig& cascade_config){
    vector<Rect>&& objects = detect_objects(
        image, detector, cascade_config
    );
    for(Rect rect : objects){
        rectangle(image, rect, Scalar(255, 0, 0));
        ostringstream oss;
        oss << rect.width;
        putText(
            image, oss.str(), rect.tl(),
            FONT_HERSHEY_PLAIN, 1, Scalar(255, 255, 255)
        );
    }
}

void run(CascadeConfig& cascade_config){
    string window_name("window");
    VideoCapture cap;
    if(cascade_config.stream.size() == 0)
        cap.open(0);
    else
        cap.open(cascade_config.stream);
    namedWindow(window_name);
    using seconds = chrono::duration<float, ratio<1,1>>;
    chrono::high_resolution_clock clock;
    auto btime = clock.now();
    Stats stats;
    auto&& detector = make_detector(cascade_config);
    while(true){
        auto duration_dt = clock.now() - btime;
        btime += duration_dt;
        float dt = seconds(duration_dt).count();
        update_stats(stats, dt);
        Mat image;
        if(!cap.read(image))
            break;
        draw_stats(stats, image);
        detect(image, detector, cascade_config);
        imshow(window_name, image);
        if(waitKey(20) == ' ')
            break;
    }
    cap.release();
    destroyWindow(window_name);
}

int main(int argc, char** argv){
    // string url;
    // if(argc != 2){
    //     url = "";
    // }else{
    //     url = argv[1];
    // }
    //setUseOptimized(false);
    setNumThreads(1);
    ocl::setUseOpenCL(false);
    auto cascade_config = config(argc, argv);
    run(cascade_config);
    return 0;
}
